posts/2024-06-20-liberty-user-feature-tutorial.adoc

```diff
@@ -0,0 +1,252 @@
+
+Then add the following method
+
+```
+	public Set<Class<?>> getBeanClasses() {
+		return Set.of(CDIProducer.class);
+	}
+```
+
+Before proceeding to the next step it is highly advised to take a look at the javadoc for ` CDIExtensionMetadata` which can be found at https://openliberty.io/docs/latest/reference/javadoc/spi/cdi-1.2.html
+
+It is also important to be aware that `getBeanClasses()` is a unique Open Liberty idiom. The normal way to add a new bean would be to make a class that implements `javax.enterprise.inject.spi.Extension` and register it via `META-INF/services`.
+
+If you wish to use `Extension` for compatibility with other Java EE servers or because your integration requires the power of a full `Extension` then `CDIExtensionMetadata` has a different method you can use for this purpose. If you want to register your extension via `META-INF/services` rather than ` CDIExtensionMetadata` see this guide here: https://openliberty.io/docs/latest/reference/feature/bells-1.0.html
+
+==== Step one B: Write the pom.xml.
+
+Open the `pom.xml` for the integration subproject.
+
+The `pom.xml` already contains all the dependencies we need to compile and an unconfigured maven-bundle-plugin. That is the next step.
+
+The bundle needs a human readable `<Bundle-Name>`, a machine readable `<Bundle-SymbolicName>`, and we need to provide a list of packages to include in the bundle.
+
+Inside `<instructions>` add the line `<Bundle-Name>example.user.feature.human.name</Bundle-Name>` and `<Bundle-SymbolicName>example.user.feature.integration.machine.name</Bundle-SymbolicName>`.
+
+Also inside `<instructions>` you will find the tag `<Export-Package>`, populate it with.
+
+```
+${new.integration.code.api.package};version="1.0.0",
+${new.integration.code.private.package};version="1.0.0"
+```
+
+These classes will not be registered correctly without a version number.
+
+Going back to the parent `pom.xml` set these properties:
+
+```
+<new.integration.code.private.package>com.ibm.example.cdi.internal</new.integration.code.private.package>    <new.integration.code.api.package>com.ibm.example.cdi.api</new.integration.code.api.package>
+```
+
+==== Step two: Create the ESA.
+
+Open Liberty features are packaged as Enterprise Subsystem Archive.  We will create one that will include both our new integration code and the pre-existing library.
+
+Open `esa/pom.xml`
+
+The first thing we need to do is ensure our ESA will have a `manifest.mf `file. Set `<generateManifest>true</generateManifest>` in the configuration of `esa-maven-plugin`.
+
+Now in instructions we will set a subystem symbolic name `<Subsystem-SymbolicName>example.user.feature.esa.machine.name;visibility:=public</Subsystem-SymbolicName>`. Seting the visibility is required.
+
+We will also need an IBM shortname. Add `<IBM-ShortName>${feature.name}</IBM-ShortName>` inside instructions, and set ${feature.name} to `example-feature` in the root pom.xml
```

Comment:
here again, i think we should show a comprehensive example of the config. In an actual guide, you have the benefit of the code pane. But here I think we need examples to reduce the cognitive load on the user of translating text sentences to code changes